### PR TITLE
Try: bundle the `WP_Theme_JSON_Resolver` class instead of inheriting per WordPress version

### PR DESCRIPTION
**Pull request by oandregal**
## What?

This PR bundles the `WP_Theme_JSON_Resolver_Gutenberg` class into a single file, continuing the work done for the `WP_Theme_JSON` class at https://github.com/WordPress/gutenberg/pull/46579

## Why?

The existing mechanism uses an inheritance model, where Gutenberg modifies only certain parts of WordPress. This approach:

- has proven buggy: we've missed many backports.
- is difficult to maintain: syncing core<=>Gutenberg is time-consuming and error-prone.
- does not allow us to do certain things, for example, _remove_ specific methods, change the signature, etc.
- we've done some hacks to clean the static variables set by core: [42756](https://github.com/WordPress/gutenberg/pull/42756) and [42776](https://github.com/WordPress/gutenberg/pull/42776).

## How?

This PR creates a single file under `lib/` for:

- `WP_Theme_JSON_Resolver_Gutenberg`
- `theme.json`
- `theme-i18n.json`

## Testing Instructions

Given this moves code from one place to another, it's difficult to provide a test plan. The fact that automated tests pass gives us some certainty the main paths are working.

This is what I've done, though you may want to try things you're most familiar with:

Theme with `theme.json` (TwentyTwentyThree):

- In the site editor, make some changes to the global styles (e.g.: select a different style variation) and verify they are applied everywhere (editor, front-end).
- In the site editor, I've added a new color to the global palette and updated an existing one.
- In the editors, verify that presets (typography, colors, etc.) are properly listed (no values missing, etc.).

Theme without `theme.json` (TwentyTwenty):

- Load the theme in the editor and front-end. Verify that it works as before.
- In the editors, verify that presets (typography, colors, etc.) are properly listed (no values missing, etc.).

## Questions

**What to do with experimental code**

Experimental code is code that is not ready for backport yet. So far, we've only run into 1 use case I know of: webfonts. For this use case, we only need to call `gutenberg_add_registered_webfonts_to_theme_json` in two places (I've marked them as comments in the PR).

In this PR, I propose we simply mark this code as experimental via comments. It's simple enough and works well for this case.

An alternative would be keeping the experimental class and override the `get_theme_data` method. This is what we did now and I'd try to avoid it. We missed a couple of things using this system (and [many others](https://github.com/WordPress/gutenberg/pull/46579#issuecomment-1354411630) in the `WP_Theme_JSON` class):

- The [theme support for `appearanceTools`](https://github.com/WordPress/gutenberg/pull/43337/files) missed the window for being backported to WordPress 6.1.
- The Gutenberg code didn't have changes from core (such as translating the parent theme of a child theme or reorganizing how default parameters work).

### CODE REVIEW

**Review by spacedmonkey — APPROVED** [review]
YES! One file to port. Let's do this for all files going forward.

File: lib/class-wp-theme-json-resolver-gutenberg.php [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+			$wp_theme        = wp_get_theme();
+			if ( '' !== $theme_json_file ) {
+				$theme_json_data = static::read_json_file( $theme_json_file );
+				$theme_json_data = static::translate( $theme_json_data, $wp_theme->get( 'TextDomain' ) );
+			} else {
+				$theme_json_data = array();
+			}
+			// BEGIN OF EXPERIMENTAL CODE. Not to backport to core.
+			$theme_json_data = gutenberg_add_registered_webfonts_to_theme_json( $theme_json_data );
+			// END OF EXPERIMENTAL CODE.
```
Comment (hellofromtonya) [inline lib/class-wp-theme-json-resolver-gutenberg.php:250]: I'm doing something similar in the (Web) Fonts API, i.e. wrapping code blocks in inline comments to alert not to backport to Core. I used:

```
BACKPORT NOTE: <explanation>
```

How to mark "don't backport" code though likely should be standardized across the codebase. For example, may want to add an annotation such as `@backportNote`, `@doNotBackport`, etc. Though this is likely best discussed in a separate issue and then added to the docs.

**Comment by spacedmonkey** [comment]
Can we do this for other classes in this repo. It could go a long way to help porting files between the repos.

**Review by scruffian — APPROVED** [review]
Tested in Loudness:
✅ Changing the background and text color
✅ Using refs in theme.json
✅ Using hover in theme.json

Also
✅ Changing the core theme.json file
✅ Changing settings in Global Styles, saving and reloading the frontend.

**Comment by scruffian** [comment]
Lets get this in ASAP to avoid having to deal with the conflicts

**Comment by hellofromtonya** [comment]
I too like the idea of bundling all changes into 1 class and file. As you note, it'll ease backporting and maintaining efforts.

On the Core side though, need to think about how to keep this file synchronized with what's in Core. Why? The `WP_Theme_JSON_Resolver` can and will change independently in Core. Since this PR no longer extends from Core's class, any changes made in Core to fix bugs will need to ported back to Gutenberg to keep both in sync. Why? 

* To avoid unexpected behaviors when testing and further developing here.
* To ease conflicts at merge into Core time.

I don't yet have an answer of how to keep things synchronized. But flagging it here for awareness and further thinking.

**Comment by oandregal** [comment]
> On the Core side though, need to think about how to keep this file synchronized with what's in Core. Why? The WP_Theme_JSON_Resolver can and will change independently in Core.

Yeah. Ideally, all changes are made in Gutenberg, _first_. Though I understand that this is a conversation larger than this PR and can also benefit/require some automation, in addition to clarify around flows.

> Since this PR no longer extends from Core's class, any changes made in Core to fix bugs will need to ported back to Gutenberg to keep both in sync. Why?

Even with the previous setup, most of the methods were overridden, so most changes in core were ignored. Not inheriting also gives us peace of mind to avoid running into unexpected conflicts with some sites running the nightlies.

**Comment by oandregal** [comment]
Tested the TwentyTwentyThree theme in a variety of scenarios: updated user data (via global styles sidebar), updated theme data (via theme.json of the theme), and update core data (theme.json bundled with gutenberg). Everything works as expected.

**Comment by oandregal** [comment]
Tested TwentyTwenty, including modifying some values from the `theme.json` bundled with Gutenberg. Works as expected.

**Comment by oandregal** [comment]
Going to merge this one as soon as tests pass.

**Comment by oandregal** [comment]
Thanks everyone for the quick turnaround on this one :)

**Comment by noahtallen** [comment]
I'm including this in a patch release of 14.8, because it is required for compatibility with WordPress 6.0. I looked through the PR changes, and each deleted file was not modified after 14.8 was released. Assuming this PR is mostly a refactor (maintaining functionality rather than changing how the feature works), including it in 14.8 won't functionally change how 14.8 works. See #46811 for more info

**Pull request merged**